src/libraries/Common/tests/System/Net/Http/Http2LoopbackConnection.cs

```diff
@@ -224,11 +228,39 @@ private async Task<Frame> ReadFrameAsync(CancellationToken cancellationToken)
                     return GoAwayFrame.ReadFrom(header, data);
                 case FrameType.Continuation:
                     return ContinuationFrame.ReadFrom(header, data);
+                case FrameType.WindowUpdate:
+                    return WindowUpdateFrame.ReadFrom(header, data);
                 default:
                     return header;
             }
         }
 
+        private async Task<bool> TryProcessExpectedPingFrameAsync(PingFrame pingFrame)
+        {
+            if (_expectPingFrame != null)
+            {
+                _expectPingFrame.SetResult(pingFrame);
+                _expectPingFrame = null;
+                return true;
+            }
+            else if (_transparentPingResponse && !pingFrame.AckFlag)
+            {
+                bool shutdownOccured = false;
+                try
+                {
+                    await SendPingAckAsync(pingFrame.Data);
+                }
+                catch (IOException ex) when (_expectClientDisconnect && ex.InnerException is SocketException se && se.SocketErrorCode == SocketError.Shutdown)
+                {
+                    // couldn't send PING ACK, because client is already disconnected
+                    shutdownOccured = true;
+                }
+                _transparentPingResponse = !shutdownOccured;
```

Comment:
Nit: seems like you could just do `_transparentPingResponse = false` in the catch and not need `shutdownOccured` at all.